web/flat-web/src/utils/CoursewarePreloader/index.ts

```diff
@@ -0,0 +1,68 @@
+import { BlobReader, BlobWriter, ZipReader } from "@zip.js/zip.js";
+import { contentType, openCache } from "./utils";
+
+class CoursewarePreloader {
+    public downloaders = new Map<string, Response>();
+
+    public async preload(pptSrc: string): Promise<void> {
+        const pptFiles = /(static|dynamic)Convert\/([0-9a-f]{32})\//.exec(pptSrc);
+        if (!pptFiles) return;
+        const [, pptType, taskUUID] = pptFiles;
+
+        if (this.downloaders.has(taskUUID)) {
+            return;
+        }
+
+        const zipUrl = `https://convertcdn.netless.link/${pptType}Convert/${taskUUID}.zip`;
+
+        const res = await fetch(zipUrl, {
+            method: "get",
+        });
+
+        this.downloaders.set(taskUUID, res);
+
+        if (res.status !== 200) {
+            if (this.downloaders.has(taskUUID)) {
+                this.downloaders.delete(taskUUID);
+            }
+            throw new Error(
+                `download task ${JSON.stringify(taskUUID)} failed with status ${res.status}`,
+            );
+        }
+
+        const resourceUrlPrefix = `https://convertcdn.netless.link/${pptType}Convert`;
+
+        this.unzip(res, resourceUrlPrefix);
+    }
+
+    private async unzip(res: Response, resourceUrlPrefix: string): Promise<void> {
+        let reader: ZipReader | undefined;
+        try {
+            reader = new ZipReader(new BlobReader(await res.blob()));
+            const entries = await reader.getEntries();
+
+            for (const entry of entries) {
+                if (!entry.getData) continue;
+                const blob: Blob = await entry.getData(new BlobWriter());
+                const resourceUrl = `${resourceUrlPrefix}/${entry.filename}`;
+                await this.cacheEntry(resourceUrl, blob);
+            }
+        } catch (err) {
+            console.log(err);
+        }
+    }
+
+    private async cacheEntry(resourceUrl: string, blob: Blob): Promise<void> {
+        const cache = await openCache("flat");
+        const response = new Response(blob, {
+            headers: { "Content-Type": contentType(resourceUrl) },
+        });
+        cache.put(resourceUrl, response);
+    }
+}
+
+let coursewarePreloader: CoursewarePreloader | undefined;
+
+export function getCoursewarePreloader(): CoursewarePreloader {
+    return (coursewarePreloader ??= new CoursewarePreloader());
+}
```

Comment:
Why not write it as:

```ts
export const coursewarePreloader = new CoursewarePreloader();
```

Reply:
lazy loading

Reply:
It will increase the readability of the code, this is unnecessary. Instantiating a new class will not cause any major performance problems.

Even this class does not have `constructor`